scenario/utils.py

```diff
@@ -0,0 +1,32 @@
+from contextlib import contextmanager
+from typing import Type
+
+from ops.framework import EventBase, Framework, CommitEvent, PreCommitEvent
+
+
+@contextmanager
+def capture_events(*types: Type[EventBase], include_lifecycle_events=False):
+    """Capture all events of type `*types` (using instance checks).
+
+    If no types are passed, will capture all types.
+    """
+    def _filter(evt):
+        if not include_lifecycle_events and isinstance(evt, (CommitEvent, PreCommitEvent)):
+            return False
+        if types:
+            return isinstance(evt, types)
+        return isinstance(evt, EventBase)
+
+    captured = []
+    _real_emit = Framework._emit  # type: ignore # noqa # ugly
```

Comment:
I realize Python has these global capture things like `assertWarns`, but it doesn't seem great to me that this patches globally (the global `Framework._emit`). Is there some way to access the `charm.framework` instead, and only patch for that instance?

Something like:

```python
with capture_events(charm, StartEvent) as captured:
    State().trigger('start', charm)
```

Or maybe the test doesn't have access to the charm instance, only the type? In that case is there a way to work in with `trigger()` to only patch a single instance's framework._emit?

Reply:
Fair enough. If (something like) this was added to `Harness`, there is a framework instance there, so we could add it as `Harness.capture_events()` or similar. Or event just always capture events in the harness and you'd access it as `emitted_events` (you could use a list comprehension with `isinstance` to filter).